sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceManager.scala

```diff
@@ -34,23 +32,29 @@ import org.apache.spark.util.Utils
  * A manager for user-defined data sources. It is used to register and lookup data sources by
  * their short names or fully qualified names.
  */
-class DataSourceManager extends Logging {
+class DataSourceManager(
+    initDataSourceBuilders: => Option[
+      Map[String, UserDefinedPythonDataSource]] = None
+   ) extends Logging {
+  import DataSourceManager._
   // Lazy to avoid being invoked during Session initialization.
   // Otherwise, it goes infinite loop, session -> Python runner -> SQLConf -> session.
-  private lazy val dataSourceBuilders = {
-    val builders = new ConcurrentHashMap[String, UserDefinedPythonDataSource]()
-    builders.putAll(DataSourceManager.initialDataSourceBuilders.asJava)
-    builders
+  private lazy val staticDataSourceBuilders = initDataSourceBuilders.getOrElse {
+    initialDataSourceBuilders
```

Comment:
I think the DataSourceManager is session-level and should not be the one to initialize static data sources. When we initialize the DataSourceManager for each spark session, we can pass in the static ones.

So it might make more sense to have an API in SparkContext for static data sources?

Reply:
Yeah I agree .. but the problem is that `UserDefinedPythonDataSourceLookupRunner.runInPython` requires `SQLConf.get` that requires `SparkSession` initialization.

So, this initialization of static datasources must happen at least when a session is created. So, I here put the static initialization logic into the first call of `DataSourceManager` in any session for now.

Reply:
Ah because of these two configs:
```
    val simplifiedTraceback: Boolean = SQLConf.get.pysparkSimplifiedTraceback
    val workerMemoryMb = SQLConf.get.pythonPlannerExecMemory
```
I think instead of accessing the SQLConf here, we should pass them as parameters to this method `runInPython` to avoid this initialization issue. Maybe we can add a TODO for a follow up PR?